src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs

```diff
@@ -220,70 +241,59 @@ private void GatherMiningStatistics()
 
             uint currentSlotTime = (uint)this.dateTimeProvider.GetAdjustedTimeAsUnixTimestamp();
             currentSlotTime -= currentSlotTime % this.network.ConsensusOptions.TargetSpacingSeconds;
-            if (currentHeader.Header.Time > currentSlotTime)
-                currentSlotTime = currentHeader.Header.Time;
-
-            // Determine the number of slots before this node will mine.
-            uint slotOffset = (this.slotsManager.GetMiningTimestamp(currentSlotTime) - currentSlotTime) / this.network.ConsensusOptions.TargetSpacingSeconds;
-
-            // Determine the current slot from that.
-            int mySlotIndex = modifiedFederation.FindIndex(m => m.PubKey == this.federationManager.CurrentFederationKey?.PubKey);
-            int currentSlot = (int)(mySlotIndex - slotOffset) % modifiedFederation.Count;
-            while (currentSlot < 0)
-                currentSlot += modifiedFederation.Count;
 
             // Determine the public key of the current slot.
-            PubKey pubKey = modifiedFederation[currentSlot].PubKey;
+            PubKey pubKey;
+            string pubKeyRepresentation;
 
             // Iterate mining slots.
-            for (int i = 0; i < maxDepth; i++)
+            for (int i = 0; i < maxDepth; i++, currentSlotTime -= this.network.ConsensusOptions.TargetSpacingSeconds)
             {
-                string pubKeyRepresentation = pubKey.ToString().Substring(0, pubKeyTakeCharacters);
-
-                if (pubKey == this.federationManager.CurrentFederationKey?.PubKey)
-                    pubKeyRepresentation = "█████";
+                while (currentHeader.Header.Time > currentSlotTime)
+                    currentHeader = currentHeader.Previous;
 
                 // Mined in this slot?
                 if (currentHeader.Header.Time == currentSlotTime)
                 {
+                    pubKey = this.federationHistory.GetFederationMemberForBlock(currentHeader)?.PubKey;
+
+                    if (pubKey == this.federationManager.CurrentFederationKey?.PubKey)
```

Comment:
Can maybe put this piece of code in a local method and just pass true or false based on `ProducedBlockInLastRound`

Reply:
Done